File: cocos2d-html5_i/IronCity/src/scene/GameScene.js
//scene: game scene.
var GameScene = cc.Scene.extend({
    moveMap:null,
    playerLayer:null,
    menuLayer:null,
    gameSceneMonster:null,
    isRectDetectedLock:false,
    laser:null,
    onEnter:function () {
        this._super();

        //map
        this.moveMap = new MovedMap();
        this.moveMap.init();
        this.moveMap.setMovedSpeed(3);
        this.addChild(this.moveMap, 0);

        //menu
        this.menuLayer = new GameSceneMenuLayer();
        this.menuLayer.init(100, "0");
        this.menuLayer.setAnchorPoint(cc.p(0, 0));
        this.menuLayer.setPosition(cc.p(0, 0));
        this.menuLayer.setScale(0.5);
        this.addChild(this.menuLayer, 0);

        //player
        this.playLayer = new GameScenePlayLayer();
        this.playLayer.init();
        this.addChild(this.playLayer, 0);

        //monster
        this.gameSceneMonster = new GameSceneMonster();
        this.gameSceneMonster.init();
        this.addChild(this.gameSceneMonster, 0);

        //laser
        this.laser = new LaserManager();
        this.laser.init();
        this.laser.scheduleUpdate();
        this.addChild(this.laser);

        //
        this.scheduleUpdate();

        //
        this.isRectDetectedLock = false;

        //this.scheduleUpdate();
        console.log("GameScene enter.111");
    },
    init:function(){
        GameScene.Scene = this;
        console.log("GameScene init.222");
    },
    //when game over.
    gameOver:function(){
        console.log("gameOver.");
        var overLayer = new GameSceneOverLayer();

        if (!overLayer.init())
        {
            overLayer = null;
            return;
        }

        this.playLayer.stopAllActions();
        this.playLayer.unscheduleUpdate();
        this.moveMap.stop();
        this.gameSceneMonster.stopAllActions();
        this.gameSceneMonster.unscheduleUpdate();

        //this.gameSceneMonster.unscheduleUpdate();

        this.menuLayer.unscheduleUpdate();

        this.addChild(overLayer, 10);
    },
    pause:function(){
        console.log("will pause.");
        this.gameSceneMonster.pause();
        console.log("will pause end.");
    },
    play:function(){
        console.log("game continue.");
        this.gameSceneMonster.play();
    },
    //tick: update all action of per-node in per-frame.
    update:function(dt){
        //console.log("update.");
        var imManArmature = this.playLayer.imManArmature;
        var actionNum = this.playLayer.actionNum;
        if(actionNum == this.playLayer.ACTION_RUN)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2+46,
                imManArmature.getPosition().y,imManArmature.getContentSize().width-90,imManArmature.getContentSize().height-50);
        }
        else if(actionNum == this.playLayer.ACTION_STAND_JUMP)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2+30,
                imManArmature.getPosition().y,imManArmature.getContentSize().width-50,imManArmature.getContentSize().height-50);
        }
        else if(actionNum == this.playLayer.ACTION_RUN_JUMP)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2+33,
                imManArmature.getPosition().y,imManArmature.getContentSize().width-70,imManArmature.getContentSize().height-50);
        }
        else if(actionNum == this.playLayer.ACTION_RUN_STOP)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2+40,
                imManArmature.getPosition().y,imManArmature.getContentSize().width-110,imManArmature.getContentSize().height-45);
        }
        else if(actionNum == this.playLayer.ACTION_RUN_ATTACK)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2,
                imManArmature.getPosition().y,imManArmature.getContentSize().width,imManArmature.getContentSize().height);

        }
        else if(actionNum == this.playLayer.ACTION_STAND_ATTACK)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2,
                imManArmature.getPosition().y,imManArmature.getContentSize().width,imManArmature.getContentSize().height);

        }
        else if(actionNum == this.playLayer.ACTION_DEATH)
        {
            this.playLayer.playerBoundingBox = cc.rect(imManArmature.getPosition().x-imManArmature.getContentSize().width/2,
                imManArmature.getPosition().y,imManArmature.getContentSize().width,imManArmature.getContentSize().height);
        }

        if(this.gameSceneMonster.MonsterIndex == MonsterType.MonsterGround_enum)
        {
            this.gameSceneMonster.MonsterAmatureBoundingBox = cc.rect(this.gameSceneMonster.MonsterAmature.getPosition().x-
                this.gameSceneMonster.MonsterAmature.getContentSize().width/2+45,this.gameSceneMonster.MonsterAmature.getPosition().y+21,
                this.gameSceneMonster.MonsterAmature.getContentSize().width-90,this.gameSceneMonster.MonsterAmature.getContentSize().height-90);
        }
        else if(this.gameSceneMonster.MonsterIndex == MonsterType.MonsterSky_enum)
        {
            this.gameSceneMonster.MonsterAmatureBoundingBox = cc.rect(this.gameSceneMonster.MonsterAmature.getPosition().x-
                this.gameSceneMonster.MonsterAmature.getContentSize().width/2+45,this.gameSceneMonster.MonsterAmature.getPosition().y+21,
                this.gameSceneMonster.MonsterAmature.getContentSize().width-90,this.gameSceneMonster.MonsterAmature.getContentSize().height-90);
        }

        if (cc.rectIntersectsRect(this.playLayer.playerBoundingBox, this.gameSceneMonster.MonsterAmatureBoundingBox))
        {
            //this.unscheduleUpdate();
            //gameSceneMonster.MonsterDestroyAction();

            var shark = CCShake.create(0.3, 10);
            this.playLayer.runAction(shark);
            this.playLayer.imManArmatureBrood-=1;
            if(this.playLayer.imManArmatureBrood<1)
            {
                GameScene.getScene().menuLayer.setBloodBarPercent(0);
                this.unscheduleUpdate();
                this.playLayer.IMDeath();
                return;
            }

            GameScene.getScene().menuLayer.setBloodBarPercent(this.playLayer.imManArmatureBrood);
        }
    }
});

GameScene.Scene = null;
GameScene.getScene = function(){
    return this.Scene;
}

File: cocos2d-x-2.2.0/projects/Ironman/Classes/GameScene.cpp
//
//  GameScene.cpp
//  Ironman
//
//  Created by Ken on 13-10-10.
//
//

#include "GameScene.h"
bool GameScene::init()
{
	gameSceneMapLayer = new GameSceneMapLayer();
	
	if(gameSceneMapLayer && gameSceneMapLayer->init())
	{
		gameSceneMapLayer->autorelease();
	}else{
		
		CC_SAFE_DELETE(gameSceneMapLayer);
	}

	gameSceneMapLayer->setMovedSpeed(3);
    gameSceneMapLayer->setPosition(ccp(0,0));
    this->addChild(gameSceneMapLayer, 0);

	menuLayer = new GameSceneMenuLayer();
	
	if(menuLayer && menuLayer->init(100, "0")){
		
		menuLayer->autorelease();
	}else{
		
		CC_SAFE_DELETE(menuLayer);
	}
    menuLayer->setAnchorPoint(ccp(0,1));
	menuLayer->ignoreAnchorPointForPosition(false);
	menuLayer->setPosition(ccp(VisibleRect::leftTop().x,VisibleRect::leftTop().y));
    this->addChild(menuLayer, 0);
    
    playLayer = new GameScenePlayLayer();
	
	if(playLayer && playLayer->init()){
		
		playLayer->autorelease();
	}else{
		
		CC_SAFE_DELETE(playLayer);
	}
	
    this->addChild(playLayer, 0);
    
	gameSceneMonster = new GameSceneMonster();
    gameSceneMonster->init();
    this->addChild(gameSceneMonster, 0);
    
    laser = LaserManager::create();
    laser->scheduleUpdate();
    this->addChild(laser);
	this->scheduleUpdate();
	isRectDetectedLock = false;
    return true;
}

static GameScene *_sharedGameScene = NULL;

GameScene* GameScene::shareGameScene()
{
    if (!_sharedGameScene) {
        _sharedGameScene = new GameScene();
        if (!_sharedGameScene->init())
        {
            CC_SAFE_DELETE(_sharedGameScene);
        }
    }
    return _sharedGameScene;
}
GameScene* GameScene::newGameScene()
{
     _sharedGameScene = new GameScene();
    if (!_sharedGameScene->init())
    {
		CC_SAFE_DELETE(_sharedGameScene);
    }
    
    return _sharedGameScene;
}
void GameScene::gameOver()
{
	GameSceneOverLayer* overLayer = new GameSceneOverLayer();
	
    if (!overLayer->init())
    {
		CC_SAFE_DELETE(overLayer);
    }
	
	playLayer->stopAllActions();
	playLayer->unscheduleUpdate();
	GameScene::gameSceneMapLayer->stop();
	gameSceneMonster->stopAllActions();
	gameSceneMonster->unscheduleUpdate();

	menuLayer->unscheduleUpdate();

	this->addChild(overLayer, 0);
}

void GameScene::update(float dt)
{
	CCArmature * imManArmature = playLayer->imManArmature;
	int actionNum = playLayer->actionNum;
	if(actionNum ==playLayer->ACTION_RUN)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2+46,imManArmature->getPosition().y,imManArmature->getContentSize().width-90,imManArmature->getContentSize().height-50);
	}
	else if(actionNum == playLayer->ACTION_STAND_JUMP)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2+30,imManArmature->getPosition().y,imManArmature->getContentSize().width-50,imManArmature->getContentSize().height-50);
	}
	else if(actionNum == playLayer->ACTION_RUN_JUMP)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2+33,imManArmature->getPosition().y,imManArmature->getContentSize().width-70,imManArmature->getContentSize().height-50);
	}
	else if(actionNum == playLayer->ACTION_RUN_STOP)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2+40,imManArmature->getPosition().y,imManArmature->getContentSize().width-110,imManArmature->getContentSize().height-45);
	}
	else if(actionNum == playLayer->ACTION_RUN_ATTACK)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2,imManArmature->getPosition().y,imManArmature->getContentSize().width,imManArmature->getContentSize().height);

	}
	else if(actionNum == playLayer->ACTION_STAND_ATTACK)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2,imManArmature->getPosition().y,imManArmature->getContentSize().width,imManArmature->getContentSize().height);

	}
	else if(actionNum == playLayer->ACTION_DEATH)
	{
		playLayer->playerBoundingBox = CCRectMake(imManArmature->getPosition().x-imManArmature->getContentSize().width/2,imManArmature->getPosition().y,imManArmature->getContentSize().width,imManArmature->getContentSize().height);
	}

	if(gameSceneMonster->MonsterIndex == MonsterGround_enum)
	{
		gameSceneMonster->MonsterAmatureBoundingBox = CCRectMake(gameSceneMonster->MonsterAmature->getPosition().x-gameSceneMonster->MonsterAmature->getContentSize().width/2+45,gameSceneMonster->MonsterAmature->getPosition().y+21,gameSceneMonster->MonsterAmature->getContentSize().width-90,gameSceneMonster->MonsterAmature->getContentSize().height-90);
	}
	else if(gameSceneMonster->MonsterIndex == MonsterSky_enum)
	{
		gameSceneMonster->MonsterAmatureBoundingBox = CCRectMake(gameSceneMonster->MonsterAmature->getPosition().x-gameSceneMonster->MonsterAmature->getContentSize().width/2+45,gameSceneMonster->MonsterAmature->getPosition().y+21,gameSceneMonster->MonsterAmature->getContentSize().width-90,gameSceneMonster->MonsterAmature->getContentSize().height-90);
	}

	if (playLayer->playerBoundingBox.intersectsRect(gameSceneMonster->MonsterAmatureBoundingBox))
	{
		//this->unscheduleUpdate();
		//gameSceneMonster->MonsterDestroyAction();
		playLayer->imManArmatureBrood-=1;
		if(playLayer->imManArmatureBrood<1)
		{
			GameScene::shareGameScene()->menuLayer->setBroodBarPercent(0);
			this->unscheduleUpdate();
			playLayer->IMDeath();
			return;
		}

		GameScene::shareGameScene()->menuLayer->setBroodBarPercent(playLayer->imManArmatureBrood);
	}
}
